core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,212 @@
+      return -1;
+    }
+
+    boolean lastBlock = currentBlockIndex + 1 == numberOfBlocks;
+    int resultBufferOffset = off;
+    int remaining = len;
+
+    sourceStream.seek(AesGcmOutputStream.PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastBlockSize : cipherBlockSize;
+      int loaded = sourceStream.read(ciphertextBlockBuffer, 0, toLoad);
+      if (loaded != toLoad) {
+        throw new IOException("Read " + loaded + " instead of " + toLoad);
+      }
+
+      // Copy nonce
+      System.arraycopy(ciphertextBlockBuffer, 0, nonce, 0, AesGcmOutputStream.GCM_NONCE_LENGTH);
+
+      byte[] aad = AesGcmOutputStream.calculateAAD(fileAadPrefix, currentBlockIndex);
+      byte[] plaintextBlock;
+      try {
+        GCMParameterSpec spec = new GCMParameterSpec(AesGcmOutputStream.GCM_TAG_LENGTH_BITS, nonce);
+        gcmCipher.init(Cipher.DECRYPT_MODE, key, spec);
+        gcmCipher.updateAAD(aad);
+
+        plaintextBlock = gcmCipher.doFinal(ciphertextBlockBuffer, AesGcmOutputStream.GCM_NONCE_LENGTH,
+                toLoad - AesGcmOutputStream.GCM_NONCE_LENGTH);
+      } catch (GeneralSecurityException e) {
+        throw new IOException("Failed to decrypt", e);
+      }
+
+      int remainingInBlock = plaintextBlock.length - currentOffsetInPlainBlock;
+      boolean finishTheBlock = remaining >= remainingInBlock;
+      int toCopy = finishTheBlock ? remainingInBlock : remaining;
+
+      System.arraycopy(plaintextBlock, currentOffsetInPlainBlock, b, resultBufferOffset, toCopy);
+      remaining -= toCopy;
+      resultBufferOffset += toCopy;
+      currentOffsetInPlainBlock += toCopy;
+      boolean endOfStream = lastBlock && finishTheBlock;
+      if (endOfStream) {
+        break;
+      }
+      if (finishTheBlock) {
+        currentBlockIndex++;
+        currentOffsetInPlainBlock = 0;
+        lastBlock = currentBlockIndex + 1 == numberOfBlocks;
+      }
+    }
+
+    plainStreamPosition += len - remaining;
+    return len - remaining;
+  }
+
+  @Override
+  public void seek(long newPos) throws IOException {
+    if (newPos >= plainStreamSize) {
+      throw new IOException("At or beyond max stream size " + plainStreamSize + ", " + newPos);
+    }
+    currentBlockIndex = Math.toIntExact(newPos / plainBlockSize);
+    currentOffsetInPlainBlock = Math.toIntExact(newPos % plainBlockSize);
+    plainStreamPosition = newPos;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return plainStreamPosition;
+  }
+
+  @Override
+  public int read() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void close() throws IOException {
+    sourceStream.close();
+  }
+
+  @Override
+  public synchronized void mark(int readlimit) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public synchronized void reset() throws IOException {
+    throw new UnsupportedOperationException();
```

Comment:
Looks like we don't have to implement it, the abstract class `InputStream` has the same logic.